___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright Pierian Data</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

For most of this course we will be loading datasets into `pandas`, and we'll seldom worry about the format that dates take. This is because the `pandas` native data type (brought over from NumPy) is more compact and runs far more efficiently than Python's built-in datetime object.<br>Still, it can't hurt to understand `datetime` objects.

# The `datetime` module
Python has built-in date, time and datetime objects available through the `datetime` module<br>
For more info on datetime visit https://docs.python.org/3/library/datetime.html

In [1]:
# Import the entire module:
import datetime

### datetime `time` objects
Values can be passed in as keyword arguments...

In [2]:
tm = datetime.time(hour=5,minute=25,second=1)
tm

datetime.time(5, 25, 1)

...or as positional arguments.

In [3]:
tm = datetime.time(5,25,1)
tm

datetime.time(5, 25, 1)

In [4]:
print(tm)

05:25:01


In [5]:
type(tm)

datetime.time

### datetime `date` objects

In [6]:
dt = datetime.date(2019,1,2)
dt

datetime.date(2019, 1, 2)

In [7]:
print(dt)

2019-01-02


In [8]:
type(dt)

datetime.date

### datetime `datetime` objects

In [9]:
d = datetime.datetime(2019, 1, 2, 5, 25, 1)
d

datetime.datetime(2019, 1, 2, 5, 25, 1)

In [10]:
print(d)

2019-01-02 05:25:01


In [11]:
type(d)

datetime.datetime

<font color=green>When no time data is provided, minimum values are used:</font>

In [12]:
d = datetime.datetime(2019, 2, 2)
print(d)

2019-02-02 00:00:00


### Selective import
For efficiency, we can import just those object classes we plan to use.

In [13]:
from datetime import datetime, date, time

d = datetime(2019, 3, 1, 15, 10)    # this is easier to type
print(d)

2019-03-01 15:10:00


## `date`, `time`, and `datetime` components
We can access specific elements of the date and time within each object.

In [14]:
print(tm)
print(tm.minute)

05:25:01
25


In [15]:
print(dt)
print(dt.day)

2019-01-02
2


In [16]:
print(d)
print(d.second)

2019-03-01 15:10:00
0


<font color=green>Of course, time objects don't contain date information, and date objects don't store time.</font>

In [17]:
print(tm.day)

AttributeError: 'datetime.time' object has no attribute 'day'

In [18]:
print(dt.second)

AttributeError: 'datetime.date' object has no attribute 'second'

## Today's date
Both `date` and `datetime` objects offer a `.today()` method that returns the current date as determined by the computer system clock.

In [19]:
x = date.today()
print(x)

2019-01-03


In [20]:
y = datetime.today()
print(y)

2019-01-03 12:15:05.526582


<font color=green>Note that assignments take a snapshot of the current date and store it. This value doesn't move forward with time.</font>

In [21]:
print(y)

2019-01-03 12:15:05.526582


## Useful methods

In [22]:
d = datetime(1969,7,20,20,17)

**`d.weekday()`** returns the day of the week as an integer, where Monday is 0 and Sunday is 6

In [23]:
d.weekday()

6

**`d.isoweekday()`** returns the day of the week as an integer, where Monday is 1 and Sunday is 7

In [24]:
d.isoweekday()

7

**`d.replace()`** returns a modified copy of the original, permitting substitutions for any date/time attribute

In [25]:
d.replace(year=1975,month=3)

datetime.datetime(1975, 3, 20, 20, 17)

<font color=green>Note that <tt>d.replace()</tt> does not change the original.</font>

In [26]:
print(d)

1969-07-20 20:17:00


## Time tuples
`datetime.timetuple()` returns a named tuple of values. Note that `date.timetuple()` returns 0 values for time elements.

In [32]:
r = date(2004,10,27)
s = datetime(2004,10,27,20,25,55)

In [33]:
r.timetuple()

time.struct_time(tm_year=2004, tm_mon=10, tm_mday=27, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=2, tm_yday=301, tm_isdst=-1)

In [34]:
s.timetuple()

time.struct_time(tm_year=2004, tm_mon=10, tm_mday=27, tm_hour=20, tm_min=25, tm_sec=55, tm_wday=2, tm_yday=301, tm_isdst=-1)

<table style="display: inline-block">
<caption style="text-align: center">**TIME TUPLE VALUES**</caption>
<tr><th>NAME</th><th>EQUIVALENT</th><th>EXAMPLES</th></tr>
<tr><td>tm_year</td><td>d.year</td><td>2004</td></tr>
<tr><td>tm_mon</td><td>d.month</td><td>10</td></tr>
<tr><td>tm_mday</td><td>d.day</td><td>27</td></tr>
<tr><td>tm_hour</td><td>d.hour</td><td>20</td></tr>
<tr><td>tm_min</td><td>d.minute</td><td>25</td></tr>
<tr><td>tm_sec</td><td>d.second</td><td>55</td></tr>
<tr><td>tm_wday</td><td>d.weekday()</td><td>2</td></tr>
<tr><td>tm_yday</td><td>see below</td><td>301</td></tr></table>

**tm_yday** is the number of days within the current year starting with 1 for January 1st, as given by the formula<br>
&emsp;&emsp; <tt>yday = d.toordinal() - date(d.year, 1, 1).toordinal() + 1</tt><br><br>
**tm_isdst** relates to timezone settings which we'll cover in an upcoming section.

### This just scratches the surface
There's a lot we can do with Python datetime objects as far as formatting their appearance, parsing incoming text with the 3rd party <tt>[dateutil](https://dateutil.readthedocs.io/en/stable/)</tt> module, and more. For now, we'll leave this alone and focus on NumPy. NumPy's `datetime64` dtype encodes dates as 64-bit integers, so that arrays of dates are stored very compactly.